src/Build.OM.UnitTests/Definition/ProjectItem_Tests.cs

```diff
@@ -2173,6 +2173,49 @@ public void RemoveWithItemReferenceOnMatchingMetadata()
             items.ElementAt(2).GetMetadataValue("M2").ShouldBe("d");
         }
 
+        [Fact]
+        public void RemoveWithItemReferenceOnMatchingMetadataReferences()
+        {
+            string content = 
+                @"<Project>
+                    <PropertyGroup>
+                        <p0>v0</p0>
+                    </PropertyGroup>
+                    <ItemGroup>
+                        <Meta2 Include='M2'/>
+                        <Meta2 Include='g'/>
+
+                        <I1 Include='a1' v0='1' M2='a'/>
+                        <I1 Include='b1' v0='2' M2='x'/>
+                        <I1 Include='c1' v0='3' M2='y'/>
+                        <I1 Include='d1' v0='4' M2='b'/>
+
+                        <I2 Include='a2' v0='x' m2='c'/>
+                        <I2 Include='b2' v0='2' m2='x'/>
+                        <I2 Include='c2' v0='3' m2='Y'/>
+                        <I2 Include='d2' v0='y' m2='d'/>
+
+                        <I2 Remove='@(I1)' MatchOnMetadata='$(Meta1);@(Meta2)' />
```

Comment:
`Meta2` also contains `g`, which is not defined as metadata on any of the items. This should lead to Remove not matching anything, right? Since all metadata need to match, a `MatchOnMetadata` where one of the metadata names does not exist should lead to non matches. Maybe create a new test with `MatchOnMetadata="a;b"`, where all items have `a`, but not all items have `b`. The ones without `b` should not get removed.

Reply:
My understanding of the desired behavior was that the metadata not existing on both items counts as it matching (unless none of the metadata is defined, in which case nothing happens). That is, here, none of I1 or I2 have g, therefore g matches on both. That's the thinking with the check [here](https://github.com/microsoft/msbuild/pull/4997/commits/e8d60e37b69c47ac19e2d114fa5b932749ad4b3c#diff-95620c6d00a356fcc2a645b390eb351bR94). Are you suggesting that, in order for it to be a match, all metadata must both exist and match?

Reply:
Good point. Let's have it like you're proposing, I can't find any arguments against it. :)
I updated the PR description with the following observation: `- if a metadata name doesn't exist in both items, then it's considered a match. Degenerate case: if none of the metadata exist in both items, it's a match,`.

We can see what other people think about it during the PR meeting.